### PR TITLE
Print entire cause chain in case of actor restart

### PR DESCRIPTION
**Pull request by thomaseizinger**
Previously, the stop reason of a supervised actor only needed to
implement `fmt::Display`. In practise however, we are always using
an enum here that implements `std::error::Error`. The default impl of
`Display` for an `Error` only prints the top-level cause. If the error
contains a `cause`, it is the responsibility of the caller to walk the
chain of causes and concatenate all the errors.

Anyhow implements this logic for its own error type and it can be used
by specifying the `alternate` modifier of `Display`: `#`.

To construct an anyhow::Error, we change the trait requirements for the
stop reason to implement `std::error::Error`.

### CODE REVIEW

**Review by klochowicz — COMMENTED** [review]
out of curiosity - how does this look like? have you managed to capture some examples is the logs? has it proven more useful?

**Comment by thomaseizinger** [comment]
> out of curiosity - how does this look like? have you managed to capture some examples is the logs? has it proven more useful?

It looks like other cause chains that we print out in the logs. They are concatenated with `:`. Something like:

```
Failed to deserialize long_offer event: invalid type: map, expected unit at line 1 column 0
```

File: xtras/src/supervisor.rs [hunk 1]
```diff
@@ -144,13 +148,13 @@ where
...
-    R: fmt::Display + Send + 'static,
+    R: Error + Send + Sync + 'static,
     S: Into<R> + Send + 'static,
 {
     pub fn handle(&mut self, msg: Stopped<R>, ctx: &mut Context<Self>) {
         let actor = T::name();
-        let reason_str = msg.reason.to_string();
-        let should_restart = (self.restart_policy)(msg.reason);
+        let should_restart = (self.restart_policy)(&msg.reason);
+        let reason_str = format!("{:#}", anyhow::Error::new(msg.reason)); // Anyhow will format the entire chain of errors when using `alternate` Display (`#`)
```
Comment (luckysori) [inline xtras/src/supervisor.rs:157]: 👍 Nice trick. It's interesting that it's not easy to print the error chain without converting to `anyhow::Error`.

**Review by luckysori — APPROVED** [review]

File: xtras/src/supervisor.rs (hunk 1 above)
Comment (thomaseizinger) [reply xtras/src/supervisor.rs:157]: It is in the works: https://github.com/rust-lang/project-error-handling/issues/10#issuecomment-720652610

**Pull request merged**